recovery_backend/backend/configmap.go

```diff
@@ -0,0 +1,219 @@
+	}
+
+	cm.Data[version] = string(dataByte)
+	_, err = rb.Datastore.UpdateConfigMap(cm)
+
+	return err
+}
+
+func (rb *RecoveryBackend) RecovRemoveClientID(configMapName, clientID string) error {
+	logrus.Infof("Remove client '%v' in recovery backend %v", clientID, configMapName)
+
+	cm, err := rb.Datastore.GetConfigMap(rb.Namespace, configMapName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get configmap %v", configMapName)
+	}
+
+	version, ok := cm.Annotations["version"]
+	if !ok {
+		return errors.Wrapf(err, "failed to get version from configmap %v", configMapName)
+	}
+
+	dataStr := cm.Data[version]
+	if dataStr == "" {
+		return errors.Wrapf(err, "failed to get data from recovery backend %v (version %v)", configMapName, version)
+	}
+
+	data := map[string][]string{}
+	if err := json.Unmarshal([]byte(dataStr), &data); err != nil {
+		return errors.Wrapf(err, "unable to decode data from recovery backend %v (version %v)", configMapName, version)
+	}
+
+	delete(data, clientID)
+
+	dataByte, err := json.Marshal(data)
+	if err != nil {
+		return errors.Wrapf(err, "unable to encode data (recovery backend=%v, version %v)", configMapName, version)
+	}
+
+	cm.Data[version] = string(dataByte)
+	_, err = rb.Datastore.UpdateConfigMap(cm)
+
+	return err
+}
+
+func (rb *RecoveryBackend) RecovReadClientIDs(configMapName string) ([]string, error) {
+	logrus.Infof("Read clients from recovery backend %v", configMapName)
+
+	cm, err := rb.Datastore.GetConfigMap(rb.Namespace, configMapName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get configmap %v", configMapName)
+	}
+
+	version, ok := cm.Annotations["version"]
+	if !ok {
+		logrus.Infof("Annotation version is not existing in recovery backend %v", configMapName)
+		return []string{}, nil
+	}
+
+	dataStr := cm.Data[version]
+	if dataStr == "" {
+		return []string{}, errors.Wrapf(err, "failed to get data from recovery backend %v (version %v)", configMapName, version)
+	}
+
+	data := map[string][]string{}
+	if err := json.Unmarshal([]byte(dataStr), &data); err != nil {
+		return []string{}, errors.Wrapf(err, "unable to decode data from recovery backend %v (version %v)", configMapName, version)
+	}
+
+	clients := []string{}
+	for c := range data {
+		clients = append(clients, c)
+	}
+
+	return clients, nil
+}
+
+func (rb *RecoveryBackend) RecovAddRevokeFilehandle(configMapName, version, clientID, revokeFilehandle string) error {
+	logrus.Infof("Add client %v revoke filehandle %v into recovery backend %v (version %v)",
+		clientID, revokeFilehandle, configMapName, version)
+
+	cm, err := rb.Datastore.GetConfigMap(rb.Namespace, configMapName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get configmap %v", configMapName)
+	}
+
+	dataStr := cm.Data[version]
+	if dataStr == "" {
+		return errors.Wrapf(err, "failed to get data from recovery backend %v (version %v)", configMapName, version)
+	}
+
+	data := map[string][]string{}
```

Comment:
Could we use type alias to make the code easy to understand?

```golang
type clientID string
type revokeFileHandle string

data := map[clientID][]revokeFileHandle
```

Reply:
Addressed. Thank you.